csrc/val_graph_visitor.cpp

```diff
@@ -149,4 +149,387 @@ void ValGraphVisitor::traverse() {
+      os << "Forward";
+      break;
+    case Direction::Backward:
+      os << "Backward";
+      break;
+    case Direction::Undefined:
+      os << "Undefined";
+      break;
+  }
+  return os;
+}
+
+std::ostream& operator<<(std::ostream& os, const ExprPath& path) {
+  for (const auto& [expr_group, direction] : path) {
+    NVF_ERROR(!expr_group->empty());
+    Expr* expr = expr_group->front();
+    NVF_ERROR(expr != nullptr);
+    os << direction << " " << expr->toString();
+  }
+  return os;
+}
+
+ExprPath ValGraphBFS::getExprsBetween(
+    const ValGraph& graph,
+    const ValGroups& from,
+    const ValGroups& to) {
+  ValGraphBFS bfs(
+      graph,
+      {from.vector().begin(), from.vector().end()},
+      {to.vector().begin(), to.vector().end()});
+
+  bfs.traverse();
+
+  return bfs.getShortestExprPath();
+}
+
+namespace {
+
+std::string toString(const ValGraphBFS::GroupType& g) {
+  if (const ExprGroup* eg = std::get_if<ExprGroup>(&g)) {
+    std::stringstream ss;
+    ss << nvfuser::toString(*eg)
+       << ", input: " << (*eg)->front()->input(0)->name();
+    return ss.str();
+  } else if (const ValGroup* vg = std::get_if<ValGroup>(&g)) {
+    return nvfuser::toString(*vg);
+  } else {
+    NVF_ERROR(false);
+  }
+}
+
+} // namespace
+
+bool ValGraphBFS::allToGroupsVisited() const {
+  return std::all_of(
+      to_groups_.begin(),
+      to_groups_.end(),
+      [&](const GroupType& group) -> bool { return isVisited(group); });
+};
+
+void ValGraphBFS::traverse() {
+  for (const auto& g : from_groups_) {
+    setVisited(g);
+    addNewNeighbors(g);
+  }
+
+  while (!allToGroupsVisited()) {
+    bool something_was_processed = false;
+    std::deque<GroupType> not_ready_;
+    while (!allToGroupsVisited() && !to_visit_.empty()) {
+      const auto g = to_visit_.front();
+      to_visit_.pop_front();
+
+      if (isVisited(g)) {
+        continue;
+      }
+
+      auto ready_direction = isReady(g);
+      if (!ready_direction.has_value()) {
+        // To stop an infinite loop, the not-ready group is not moved
+        // back to the to_visit_ queue but kept in the separate
+        // queue. This way, if all groups in to_visit_ are not ready,
+        // the queue would eventually become empty, which would then
+        // break the inner while loop. The something_was_processed
+        // flag is used to remember if there's any progress.
+        not_ready_.emplace_back(g);
+        continue;
+      }
+
+      // Visit this group and add its neighbors to to_visit if not
+      // visited yet
+      setVisited(g);
+      setPrevGroups(g, *ready_direction);
+      addNewNeighbors(g);
+      something_was_processed = true;
+    }
+
+    // If nothing was processed, break out of the loop
+    if (!something_was_processed) {
+      break;
+    }
+
+    // Something was processed. Redo the traversal.
+    to_visit_.insert(to_visit_.end(), not_ready_.begin(), not_ready_.end());
+  };
+
+  if (!allToGroupsVisited()) {
```

Comment:
`not_ready_` is a temporary container to hold groups that are adjacent to the frontier groups but we are not yet able to visit due to missing dependencies. For example, if a `Merge` group only has one of the inputs, it would be in `not_ready_`. Those groups are connected with missing dependencies.

If `not_ready_` is empty, that should mean there's no more connected group that isn't visited yet. So, some of the `to_groups_` must be disconnected.

I'm not sure if this could be a problem with cycles. At least, it should be detected and result in a failure rather than just resulting in an infinite loop.